src/base/Disambiguate.ml

```diff
@@ -0,0 +1,888 @@
+              remove_local_id_from_dict dicts.simp_var_dict (as_string x);
+          }
+        in
+        pure @@ (PostDisSyntax.LibVar (dis_x, dis_topt, dis_e), res_dicts)
+    | LibTyp (tname, ctrs) ->
+        let%bind dis_tname = local_id_as_global_name tname in
+        let%bind dis_ctrs =
+          mapM ctrs ~f:(fun ctr ->
+              let { cname; c_arg_types } = ctr in
+              let%bind dis_cname = local_id_as_global_name cname in
+              let%bind dis_c_arg_types =
+                mapM c_arg_types
+                  ~f:(disambiguate_type dicts.ns_dict dicts.simp_typ_dict)
+              in
+              pure
+                {
+                  PostDisSyntax.cname = dis_cname;
+                  PostDisSyntax.c_arg_types = dis_c_arg_types;
+                })
+        in
+        (* tname is now in scope as a local type, and ctrs are in scope as local constructors,
+           so remove from dictionaries *)
+        let res_typ_dict =
+          remove_local_id_from_dict dicts.simp_typ_dict (as_string tname)
+        in
+        let res_ctr_dict =
+          List.fold_left ctrs ~init:dicts.simp_ctr_dict ~f:(fun dict ctr ->
+              remove_local_id_from_dict dict (as_string ctr.cname))
+        in
+        let res_dicts =
+          {
+            dicts with
+            simp_typ_dict = res_typ_dict;
+            simp_ctr_dict = res_ctr_dict;
+          }
+        in
+        pure @@ (PostDisSyntax.LibTyp (dis_tname, dis_ctrs), res_dicts)
+
+  let disambiguate_library (dicts : name_dicts) lib =
+    let { lname; lentries } = lib in
+    let%bind dis_lname = local_id_as_global_name lname in
+    let%bind dis_lentries_rev, _ =
+      foldM lentries ~init:([], dicts)
+        ~f:(fun (dis_lentries_acc_rev, dicts_acc) lentry ->
+          let%bind dis_lentry, new_dicts =
+            disambiguate_lib_entry dicts_acc lentry
+          in
+          pure (dis_lentry :: dis_lentries_acc_rev, new_dicts))
+    in
+    (* Toplevel names defined in the library are potentially in scope somewhere,
+       so must be collected and returned *)
+    let lib_simp_vars_rev, lib_typs_rev, lib_ctrs_rev =
+      List.fold_left lentries ~init:([], [], [])
+        ~f:(fun (vars_acc, typs_acc, ctrs_acc) lentry ->
+          match lentry with
+          | LibVar (x, _, _) -> (as_string x :: vars_acc, typs_acc, ctrs_acc)
+          | LibTyp (tname, ctrs) ->
+              let new_ctrs =
+                List.fold_left ctrs ~init:ctrs_acc ~f:(fun ctrs_acc' ctr ->
+                    as_string ctr.cname :: ctrs_acc')
+              in
+              (vars_acc, as_string tname :: typs_acc, new_ctrs))
+    in
+    let reverse_remove_duplicates l =
+      let rec recurser seen rest =
+        match rest with
+        | [] -> seen
+        | x :: xs ->
+            let new_seen =
+              x :: List.filter seen ~f:(fun v -> not String.(v = x))
+            in
+            recurser new_seen xs
+      in
+      recurser [] l
```

Comment:
Perhaps this should be parameterized with `~equal : 'a -> 'a -> bool` parameter and moved into our local library. Another option would be to use Core_kernel's `dedup_and_sort` function composed (although, I'm not sure at the moment if the order is essential here; should we document this?)

Reply:
Actually, the order shouldn't matter, as long as `List.Assoc.add` removes the previous entry with the same key, which I assume it does - if not, why would it need an `~equal` parameters?

Reply:
Indeed, `List.Assoc.add` removes the previous entry with the same key.

Reply:
Ah, no, this doesn't work: These are not association lists - these are lists of defined names.

If a library contains multiple definitions of the same toplevel name, then the last definition shadows the former ones. Therefore, the order does matter as long as there are duplicates. Also, the sort order is the order in which they appear in the list.

However, the lists are only used to remove the names from the contract's own library from the imported name dictionaries (simulating that the contract's own definitions shadow the imported names), so there is no point in neither sorting nor removing duplicates.

I'll just remove the function.